functions.php

```diff
@@ -7,20 +7,110 @@
 function databaseConnect(): PDO {
     $db = new PDO('mysql:host=db; dbname=collection_project', 'root', 'password');
     $db->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
+    $db->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_WARNING);
     return $db;
 }
+
 /**
  * Creates query fetching all data from the database
  *
  * @param PDO created by connection with database
  * @return array returns all of the data requested from the database
  */
-function databaseFetchAll($db): array{
-    $query = $db->prepare("SELECT `charname`, `class`, `level`, `strength`, `dexterity`, `constitution`, `intelligence`, `wisdom`, `charisma` FROM `characters`;");
+function databaseFetchAll(PDO $db): array {
+    $query = $db->prepare("SELECT `charname`, `class`, `level`, `strength`, `dexterity`, `constitution`, `intelligence`, `wisdom`, `charisma`, `id` FROM `characters` WHERE `deleted` = 0;");
     $query->execute();
     return $query->fetchAll();
 }
 
+/**
+ * Fetch character data from database so that it can be edited
+ *
+ * @param PDO database PDO
+ * @param string id of the character in the database
+ * @return array Returns all the information of the character that is to be edited
+ */
+function databaseFetchEditChar(PDO $db,string $id): array {
```

Comment:
$id here should be an int

Reply:
I tested whether or not the edit error messages work in cases where a user might have messed with your form - I changed the ID in the hidden input to be 200, and this function threw a type error - it turns out that if you pass in an ID that doesn't exist, this function tries to return false - I would suggest removing the return type hint here to allow the function to return false in these cases.

Then you can add some logic when you call databaseFetchEditChar to check to see if it's false (which indicates someone messed with your form), at which point you can redirect them somewhere with an error message

Sorry I didn't spot this when I reviewed story2!

Reply:
I think I have amended this. I have added an if statement into the char_edit page